### PR TITLE
change gossip dns conn limit by ENV

### PR DESCRIPTION
**Pull request by yancl**
so that we can run more nodes than 64 :)

### CODE REVIEW

File: protokube/pkg/gossip/mesh/gossip.go [hunk 1]
```diff
@@ -39,10 +40,23 @@ type MeshGossiper struct {
 }
 
 func NewMeshGossiper(listen string, channelName string, nodeName string, password []byte, seeds gossip.SeedProvider) (*MeshGossiper, error) {
+
+	connLimit := 64
+	gossipDnsConnLimit := os.Getenv("GOSSIP_DNS_CONN_LIMIT")
+	if gossipDnsConnLimit != "" {
+		limit, err := strconv.Atoi(gossipDnsConnLimit)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse env GOSSIP_DNS_CONN_LIMIT value: %v", gossipDnsConnLimit)
```
Comment (andrewsykim) [inline protokube/pkg/gossip/mesh/gossip.go]: also add `err` to message

File: protokube/pkg/gossip/mesh/gossip.go [hunk 2]
```diff
@@ -39,10 +40,23 @@ type MeshGossiper struct {
...
+		}
+		connLimit = limit
```
Comment (andrewsykim) [inline protokube/pkg/gossip/mesh/gossip.go]: This line can just be above,
```
connLimit, err := strconv.Atoi(gossipDnsConnLimit)
```

File: protokube/pkg/gossip/mesh/gossip.go (hunk 1 above)
Comment (yancl) [reply protokube/pkg/gossip/mesh/gossip.go]: ok, thanks.

File: protokube/pkg/gossip/mesh/gossip.go (hunk 2 above)
Comment (yancl) [reply protokube/pkg/gossip/mesh/gossip.go]: thanks:)

**Comment by yancl** [comment]
Yes, i really have considered make a PR that just changes the default conn limit from 64 to some larger number, for example, 300, and i prefer that:)

I think the problem with a lot of nodes using gossip dns maybe related to the  [issue](https://github.com/weaveworks/weave/issues/1621).

**Comment by chrislovecnm** [comment]
@yancl thanks for your patience.  I pinged @justinsb on this.  It would be good to find out why we limited it

**Comment by justinsb** [comment]
1000 would still be a limit :-)

I thought it was number of peers, I didn't realize it was anything more than that.  Do we know that this limit is in fact a problem?  I'd be inclined to just bump it to 200 - I think that's what weave did.

**Comment by chrislovecnm** [comment]
@yanci what have you tested?

**Comment by yancl** [comment]
@chrislovecnm  sorry for late response:)

Yes, we have setup a k8s cluster with gossip dns enabled. but as we start more nodes than the limit(64), the new nodes doesn't show in `kubectl get nodes`, and we found `->[xxx] error during connection attempt:xxx` in the syslog. the codes are [here](https://github.com/weaveworks/mesh/blob/0c91e692ee9e89b956b9c4dfc5feb59a0ff87699/connection_maker.go#L373).

So i think the num of the **total nodes** of a k8s cluster **can't** exceed `ConnLimit + 1` when using gossip mesh, we can make such check with the following steps:

* support the ConnLimit is 1
* and we first start 2 nodes, they mesh well.
* then we start the 3rd node, it can't connect any one of the 2 nodes above according code [here](https://github.com/weaveworks/mesh/blob/0c91e692ee9e89b956b9c4dfc5feb59a0ff87699/local_peer.go#L249)
* if the 3rd node can not make connection with any of those 2 nodes, then it can not get information from them and update the apiserver&etcd servers addrs to the `/etc/hosts` , so the node will not join the cluster.

Is that correct?

@justinsb  Yes, 1000 would still be a limit, we can bump it to 200. 64 is too small for a medium k8s cluster:)

thanks.

**Comment by levilovelock** [comment]
I ran into this issue today! 65 nodes was the maximum! Is there any way we can bypass this at the moment? How would one pass this arg through to new nodes too?

**Comment by sak-2** [comment]
When building this pr myself, k8s recognized more than 64 nodes.
I put the following settings in master nodes.
```
/etc/environment
GOSSIP_DNS_CONN_LIMIT=100
```
and I executed the following commands.
```
$ cd /var/cache/kubernetes-install
$ ./nodeup --conf=/var/cache/kubernetes-install/kube_env.yaml --v=8
```

**Comment by jaredallard** [comment]
LGTM, this has impacted us as well. Little insane that this is an issue :(

**Comment by justinsb** [comment]
So I don't think this should be an env var that has to be set.  I'm going to merge this PR, but I'm then going to send a PR that tries to set it automatically.  I'd like to remove the env var at least from having to be set every time you call kops.

My understanding is that 64 nodes shouldn't be a hard limit, but rather each node will limit itself to 64 peers.  As long as we don't form two disjoint segments, we should be able to go to much bigger sizes.

But I'll test it out!

/approve
/lgtm

**Comment by shrinandj** [comment]
I understand this issue may be long forgotten. But it might help others who land here... hence this question.

If the connection limit has been removed entirely, the original problem this PR (or having the GOSSIP_DNS_CONN_LIMIT env variable) was trying to fix is also fixed. Explicitly setting GOSSIP_DNS_CONN_LIMIT to some value less than the total number of nodes could lead to problems (of running into a clique), right?

**Pull request merged**